src/Collection.php

```diff
@@ -882,6 +882,15 @@ public function insertMany(array $documents, array $options = [])
      */
     public function insertOne($document, array $options = [])
     {
+        if (isset($options['session']) && $options['session']->isInTransaction()) {
+            if ($this->writeConcern instanceof WriteConcern && !$this->writeConcern->isDefault()) {
+                throw new InvalidArgumentException("Cannot use an inherited writeConcern during a transaction");
+            }
+            if (isset($options['writeConcern']) && !$options['writeConcern']->isDefault()) {
+                throw new InvalidArgumentException("Cannot specify 'writeConcern' option during a transaction");
```

Comment:
My understanding is that only `insertOne()` needs the "silently ignore" logic. If it sees a session option and that session is in a transaction, it should not inherit the Collection's RC/WC. Suppose we have the following in the Collection helper:

```php
$inTransaction = isset($options['session']) && $options['session'] instanceof Session && $options['session']->isInTransaction();

if ( ! isset($options['writeConcern']) && ! $inTransaction) {
    $options['writeConcern'] = $this->writeConcern;
}
```

Later, in the Operation's `execute()` method:

```php
// Constructor would already have validated the session's type (if set) by this point
$inTransaction = isset($options['session']) && $options['session']->isInTransaction();

// Constructor would already have unset a default WC by this point
if (isset($this->options['writeConcern']) && $inTransaction) {
    throw UnsupportedException::writeConcernNotSupportedInTransaction();
}
```

Under normal operation, the session's state should not change between the Collection helper deciding it's safe to inherit and `execute()` being called, since that all happens within the helper method.

I can think of two edge cases that would still allow for an error. Once SDAM monitoring is implemented, it's possible that a listener method could be invoked when the helper calls `Manager::selectServer()` (between inheritance logic and `execute()`). The session could also be altered during the helper method's execution via a thread. These cases would either require to be doing something quite strange or acting in bad faith, so I don't think we need to worry about it.

Reply:
Changed.